tests/phpunit/tests/user/query.php

```diff
@@ -1725,4 +1728,182 @@ public static function filter_users_pre_query( $posts, $query ) {
+			// Hence using get_role_caps() instead of has_cap().
+			$role_caps = $user->get_role_caps();
+			$this->assertArrayHasKey( 'install_plugins', $role_caps );
+			$this->assertTrue( $role_caps['install_plugins'] );
+		}
+	}
+
+	/**
+	 * @ticket 16841
+	 */
+	public function test_get_single_capability_invalid() {
+		$wp_user_search = new WP_User_Query( array( 'capability' => 'foo_bar' ) );
+		$users          = $wp_user_search->get_results();
+
+		$this->assertEmpty( $users );
+	}
+
+	/**
+	 * @ticket 16841
+	 */
+	public function test_get_single_capability_by_array() {
+		$wp_user_search = new WP_User_Query( array( 'capability' => array( 'install_plugins' ) ) );
+		$users          = $wp_user_search->get_results();
+
+		$this->assertNotEmpty( $users );
+		foreach ( $users as $user ) {
+			// User has the capability, but on Multisite they would also need to be a super admin.
+			// Hence using get_role_caps() instead of has_cap().
+			$role_caps = $user->get_role_caps();
+			$this->assertArrayHasKey( 'install_plugins', $role_caps );
+			$this->assertTrue( $role_caps['install_plugins'] );
+		}
+	}
+
+	/**
+	 * @ticket 16841
+	 */
+	public function test_get_single_capability_added_to_user() {
+		$user = self::factory()->user->create_and_get( array( 'role' => 'subscriber' ) );
+		$user->add_cap( 'custom_cap' );
+
+		$wp_user_search = new WP_User_Query( array( 'capability' => 'custom_cap' ) );
+		$users          = $wp_user_search->get_results();
+
+		$this->assertCount( 1, $users );
+		$this->assertSame( $users[0]->ID, $user->ID );
+		foreach ( $users as $user ) {
+			$this->assertTrue( $user->has_cap( 'custom_cap' ) );
+		}
+	}
+
+	/**
+	 * @ticket 16841
+	 */
+	public function test_get_multiple_capabilities_should_only_match_users_who_have_each_capability_test() {
+		wp_roles()->add_role( 'role_1', 'Role 1', array( 'role_1_cap' => true ) );
+		wp_roles()->add_role( 'role_2', 'Role 2', array( 'role_2_cap' => true ) );
+
+		$user = self::factory()->user->create_and_get( array( 'role' => 'role_1' ) );
+		$user->add_role( 'role_2' );
+
+		$wp_user_search = new WP_User_Query( array( 'capability' => array( 'role_1_cap', 'role_2_cap' ) ) );
+		$users          = $wp_user_search->get_results();
+
+		$this->assertCount( 1, $users );
+		$this->assertSame( $users[0]->ID, $user->ID );
+		foreach ( $users as $user ) {
+			$this->assertTrue( $user->has_cap( 'role_1_cap' ) );
+			$this->assertTrue( $user->has_cap( 'role_2_cap' ) );
+		}
+	}
+
+	/**
+	 * @ticket 16841
+	 */
+	public function test_get_multiple_capabilities_should_only_match_users_who_have_each_capability_added_to_user() {
+		$user = self::factory()->user->create_and_get( array( 'role' => 'administrator' ) );
+		$user->add_cap( 'custom_cap' );
+
+		$wp_user_search = new WP_User_Query( array( 'capability' => array( 'manage_options', 'custom_cap' ) ) );
+		$users          = $wp_user_search->get_results();
+
+		$this->assertNotEmpty( $users );
```

Comment:
Unsure of best approach but it would be good to check the number of users is correct too. 

Nit, non-blocker.

Reply:
I had such an assertion originally, but IIRC the number just changes too much if other tests create users as well. But I can double check again.